import logging

import pyarrow as pa
from pyarrow import Schema
import pyarrow.parquet as pq

_common: Schema = pa.schema([
    pa.field('case_id', pa.int32(), nullable=False),
    pa.field('locus', pa.string(), nullable=False),
    pa.field('locus_hash', pa.string(), nullable=False),
    pa.field('chromosome', pa.string(), nullable=False),
    pa.field('start', pa.int32(), nullable=False),
    pa.field('end', pa.int32(), nullable=True),
    pa.field('reference', pa.string(), nullable=False),
    pa.field('alternate', pa.string(), nullable=False),
])

consequence_schema: Schema = pa.schema([
    pa.field('variant_class', pa.string(), nullable=True),
    pa.field('hgvsg', pa.string(), nullable=True),
    pa.field('hgvsp', pa.string(), nullable=True),
    pa.field('hgvsc', pa.string(), nullable=True),
    pa.field('symbol', pa.string(), nullable=True),
    pa.field('transcript_id', pa.string(), nullable=True),
    pa.field('source', pa.string(), nullable=True),
    pa.field('biotype', pa.string(), nullable=True),
    pa.field('strand', pa.string(), nullable=True),
    pa.field('exon', pa.struct([
        pa.field('rank', pa.string()),
        pa.field('total', pa.string()),
    ]), nullable=True),
    pa.field('vep_impact', pa.string(), nullable=True),
    pa.field('consequences', pa.list_(pa.string()), nullable=True),
    pa.field('mane_select', pa.string(), nullable=True),
    pa.field('is_mane_select', pa.bool_(), nullable=False),
    pa.field('is_mane_plus', pa.bool_(), nullable=False),
    pa.field('is_picked', pa.bool_(), nullable=False),
    pa.field('is_canonical', pa.bool_(), nullable=False),
    pa.field('aa_change', pa.string(), nullable=True),
    pa.field('dna_change', pa.string(), nullable=True),
    pa.field('impact_score', pa.int32(), nullable=False),
])

occurance_schema: Schema = pa.schema([
    pa.field('part', pa.int32(), nullable=False),
    pa.field('seq_id', pa.int32(), nullable=False),
    pa.field('aliquot', pa.string(), nullable=False),
    pa.field('dp', pa.int32(), nullable=True),
    pa.field('gq', pa.int32(), nullable=True),
    pa.field('calls', pa.list_(pa.int32()), nullable=True),
    pa.field('has_alt', pa.bool_(), nullable=True),
    pa.field('quality', pa.float32(), nullable=True),
    pa.field('filter', pa.string(), nullable=True),
    pa.field('info_old_record', pa.string(), nullable=True),
    pa.field('info_baseq_rank_sum', pa.float32(), nullable=True),
    pa.field('info_excess_het', pa.float32(), nullable=True),
    pa.field('info_fs', pa.float32(), nullable=True),
    pa.field('info_ds', pa.bool_(), nullable=True),
    pa.field('info_fraction_informative_reads', pa.float32(), nullable=True),
    pa.field('info_inbreed_coeff', pa.float32(), nullable=True),
    pa.field('info_mleac', pa.int32(), nullable=True),
    pa.field('info_mleaf', pa.float32(), nullable=True),
    pa.field('info_mq', pa.float32(), nullable=True),
    pa.field('info_m_qrank_sum', pa.float32(), nullable=True),
    pa.field('info_qd', pa.float32(), nullable=True),
    pa.field('info_r2_5p_bias', pa.float32(), nullable=True),
    pa.field('info_read_pos_rank_sum', pa.float32(), nullable=True),
    pa.field('info_sor', pa.float32(), nullable=True),
    pa.field('info_vqslod', pa.float32(), nullable=True),
    pa.field('info_culprit', pa.string(), nullable=True),
    pa.field('info_dp', pa.float32(), nullable=True),
    pa.field('info_haplotype_score', pa.float32(), nullable=True),
    pa.field('zygosity', pa.string(), nullable=True),
    pa.field('ad_ref', pa.int32(), nullable=True),
    pa.field('ad_alt', pa.float32(), nullable=True),
    pa.field('ad_total', pa.int32(), nullable=True),
    pa.field('ad_ratio', pa.int32(), nullable=True),
    pa.field('phased', pa.bool_(), nullable=False),
    pa.field('parental_origin', pa.string(), nullable=True),
    pa.field('father_dp', pa.int32(), nullable=True),
    pa.field('father_gq', pa.int32(), nullable=True),
    pa.field('father_ad_ref', pa.int32(), nullable=True),
    pa.field('father_ad_alt', pa.int32(), nullable=True),
    pa.field('father_ad_total', pa.int32(), nullable=True),
    pa.field('father_ad_ratio', pa.float32(), nullable=True),
    pa.field('father_calls', pa.list_(pa.int32()), nullable=True),
    pa.field('father_zygosity', pa.string(), nullable=True),
    pa.field('mother_dp', pa.int32(), nullable=True),
    pa.field('mother_gq', pa.int32(), nullable=True),
    pa.field('mother_ad_ref', pa.int32(), nullable=True),
    pa.field('mother_ad_alt', pa.int32(), nullable=True),
    pa.field('mother_ad_total', pa.int32(), nullable=True),
    pa.field('mother_ad_ratio', pa.float32(), nullable=True),
    pa.field('mother_calls', pa.list_(pa.int32()), nullable=True),
    pa.field('mother_zygosity', pa.string(), nullable=True),
    pa.field('transmission_mode', pa.string(), nullable=True),
    pa.field('task_id', pa.int32(), nullable=False),
])

variant_schema: Schema = pa.schema([
    pa.field('variant_class', pa.string(), nullable=True),
    pa.field('symbol', pa.string(), nullable=True),
    pa.field('consequences', pa.list_(pa.string()), nullable=True),
    pa.field('vep_impact', pa.string(), nullable=True),
    pa.field('impact_score', pa.int32(), nullable=True),
    pa.field('mane_select', pa.string(), nullable=True),
    pa.field('is_mane_select', pa.bool_(), nullable=True),
    pa.field('is_mane_plus', pa.bool_(), nullable=True),
    pa.field('is_canonical', pa.bool_(), nullable=True),
    pa.field('rsnumber', pa.list_(pa.string()), nullable=True),
    pa.field('hgvsg', pa.string(), nullable=True),
    pa.field('hgvsc', pa.string(), nullable=True),
    pa.field('hgvsp', pa.string(), nullable=True),
    pa.field('dna_change', pa.string(), nullable=True),
    pa.field('aa_change', pa.string(), nullable=True),
    pa.field('transcript_id', pa.string(), nullable=True),
])

consequence_schema = pa.unify_schemas([consequence_schema, _common])
occurance_schema = pa.unify_schemas([occurance_schema, _common])
variant_schema = pa.unify_schemas([variant_schema, _common])